src/query/ts/m3db/async_encoded_step_iterator.go

```diff
@@ -0,0 +1,274 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/ts/m3db/consolidators"
+	xsync "github.com/m3db/m3x/sync"
+)
+
+type encodedStepIterAsync struct {
+	mu            sync.RWMutex
+	lastBlock     bool
+	started       bool
+	err           error
+	currentTime   time.Time
+	currentValues []float64
+	bounds        models.Bounds
+	meta          block.Metadata
+	seriesMeta    []block.SeriesMeta
+	seriesIters   []encoding.SeriesIterator
+	seriesPeek    []peekValue
+	consolidator  *consolidators.StepLookbackConsolidator
+	workerPools   xsync.PooledWorkerPool
+}
+
+func (b *encodedBlock) asyncStepIter() block.AsyncStepIter {
+	cs := b.consolidation
+	consolidator := consolidators.NewStepLookbackConsolidator(
+		time.Minute,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+		cs.consolidationFn,
+	)
+	return &encodedStepIterAsync{
+		lastBlock:    b.lastBlock,
+		currentTime:  cs.currentTime,
+		bounds:       cs.bounds,
+		meta:         b.meta,
+		seriesMeta:   b.seriesMetas,
+		seriesIters:  b.seriesBlockIterators,
+		consolidator: consolidator,
+		workerPools:  b.workerPools,
+	}
+}
+
+func (it *encodedStepIterAsync) Err() error {
+	return it.err
+}
+
+func (it *encodedStepIterAsync) Current() <-chan (block.Step) {
+	it.mu.RLock()
+	defer it.mu.RUnlock()
+	currentTime := it.currentTime
+	if !it.started || currentTime.After(it.bounds.End()) {
+		panic("out of bounds")
+	}
+
+	// if it.err != nil {
+	// 	return nil
+	// }
+
+	step := &encodedStep{
+		time:   currentTime,
+		values: it.currentValues,
+	}
+
+	ch := make(chan (block.Step))
+	it.workerPools.Go(func() {
+		// go func() {
+		ch <- step
+		// }()
+	})
+	return ch
+}
+
+func (it *encodedStepIterAsync) ValuesChannel() <-chan (block.IndexedStep) {
+	if it.err != nil {
+		return nil
+	}
+	stepCount := it.StepCount()
+	ch := make(chan (block.IndexedStep), stepCount)
+	workerPools := it.workerPools
+	var wg sync.WaitGroup
+	wg.Add(it.StepCount())
+	for i := 0; it.Next(); i++ {
+		idx := i
+		currentTime := it.currentTime
+		currentValues := it.currentValues
+		workerPools.Go(func() {
```

Comment:
Currently it does, as soon as some of the blocks will need to start blocking, this will change. Idea is to build up from the simple case of being able to use async blocks in a synchronous-ish fashion to one where some steps are blocked until additional data comes in.

Making this async at the moment mirrors the potential for out-of-order writes, to ensure that the output can be put back together again correctly on the function side

Reply:
Ah ok cool, fair enough. If the only purpose of this implementation is testing, let's move it to a test maybe?